languages/ruby/exercises/concept/basics/.docs/introduction.md

```diff
@@ -0,0 +1,55 @@
+Ruby is a dynamic [object-oriented language]. Everything in Ruby is an [object][object].
+
+There are two ways to assign objects to names in Ruby - assigning variables or constants. Variables always start with lower-case letters and use [snake case][snake-case] for their formatting. A variable can be redefined to different objects over its lifetime. For example, `my_first_variable` can be defined and redefined many times using the `=` operator:
+
+```ruby
+my_first_variable = 1
+my_first_variable = "Some string"
+my_first_variable = SomeComplexObject.new
+```
+
+Constants, however, can only be assigned once. They must start with capital letters and are normally written in block capitals with words separated by underscores. For example:
+
+```ruby
+MY_FIRST_CONSTANT = 10
+
+# Redefining not allowed
+# MY_FIRST_CONSTANT = "Some String"
+```
+
+Ruby is organised into classes. Classes are defined using the `class` keyword followed by the name of the class. Classes are initialized using the `.new` method.
+
+```ruby
+# Define the class
+class Calculator
+  #...
+end
+
+# Create an instance of it and assign it to a variable
+my_first_calc = Calculator.new
+```
+
+A function within a class is referred to as a _method_. Each method can have zero or more parameters. Objects are returned from methods using the `return` keyword. If no return value is specified, the final object in the method is returned instead. Methods can also have named parameters which are defined and called using the `:` syntax.  Methods are invoked using `.` syntax.
```

Comment:
```suggestion
What may be called a _function_ in other languages are referred to as a _method_. Each method may have zero or more arguments. Objects may be returned from methods using the `return` keyword. If no return is specified, the result of the last evaluated statement of the method is returned instead. Methods can also have named parameters which are defined and called using the `:` syntax.  Methods are invoked using `.` syntax.
```

Reply:
The "Methods can also have named parameters" section should be its own paragraph, where we can talk about artiy, and method signatures.  I.E., positional arguments, default arguments, keyword arguments, etc.

Reply:
I disagree here. The exercise is not intended to cover everything there is to know about parameters and methods, but just to introduce some basics. To me, this excludes the things you're mentioning, which should IMHO be explained in a separate exercise.

Reply:
Yeah, I agree with Erik. We'll have another exercise that goes into the multitude of different ways to define methods and arguments. For this first exercise, I think showing the two most common approaches is enough I think, as the primary aim of the exercise is just to give students knowledge so that future exercises to be able to define methods using these ways.

Reply:
I agree with Erik as well, in that that regard. Yes, its own paragraph, but perhaps not in this document.  The real problem I have is referring to the argument list as "parameters".

Reply:
@kotp OK - I'll reword to take that into account - thanks. Before I do though, can you explain why you think it should be `may` not `can`? The aim is to describe what is possible (e,g, it is possible to return from a method) rather than implying that someone has permission to do something, so I feel like "can" is grammatically correct here.

Reply:
Because you may return objects using the `return` but it is not required.  Ruby `will` return an object, but you may do so either explicitly or implicitly.  (The word "can" implies that you are allowed to, while that isn't quite applicable here, you may do so, and loosely, yes, you "can" but you will do so, but you may do so in various ways.)

Does that make sense?

Reply:
I thought about it a bit, and have written up this to hopefully clarify what my thoughts are about the words Methods and Functions.  While giving an example of an article that supports this, and noting that the article also talks about arguments and parameters in a way that I don't necessarily feel fits here as well.  (We can agree to use loose terms when needed, after all.)

---

What may be called a _function_ in other languages are referred to as a
_method_.<sup>[1]</sup>

Each method may have zero or more arguments. A Ruby method is allowed to
return one object.  We may return an object from a method by being
explicit by using the keyword `return`.  This is generally reserved for
when we need to return early from a method.  If no return is specified,
then the result of the last evaluated statement in the method will be
returned.  We call this an implicit return.

In Ruby, there really are no functions, if we consider that functions
are effected only by input, by the argument that is given.  However, we
do treat some methods similarly.  An example of this would be the
apparent use of `puts` or `print` in a void context.  Those methods will
return `nil` and appears to only be effected by the input to those
methods, effecting the output to a screen for example.

Methods are generally thought of to operate based on context, related to
the object on which they are called.

Note: While I agree with the "Methods Vs. Functions" portion of the
article, I don't agree with their definition of "Arguments Vs.
Parameters" in the strictest of senses, though I agree with the idea,
and understand when folks do reference them in the described way.
Sometimes we are a bit loose with our language when talking about
things, and I get that.  We tend to understand.  I try to be careful
while instructing though, since while instructing we tend to want to be
clear, while allowing for some "white lies" for a little bit of wiggle
room, to allow for understanding based on prior knowledge.

[1]: Methods Vs. Functions https://www.ombulabs.com/blog/ruby/learning/method-s-arguments-pt-1.html

---

I hope that gives some explanation why I reviewed this portion the way I did.

Reply:
Thanks @kotp. I appreciate the notes.

Re the "can" vs "may" I think in this situation we should use "can" as we are explaining what is possible, rather than what a user may choose to do. I understand that either are valid depending on how the sentence is aimed to be interpreted though.

How is this for a rewrite?

```suggestion
Units of functionality are encapsulated in methods - similar to _functions_ in other languages. A method can optionally be defined with positional arguments, or keyword arguments that are defined and called using the `:` syntax.  Methods either implicitly return the result of the last evaluated statement, or can explicitly return an object via the `return` keyword.  Methods are invoked using `.` syntax.
```

Reply:
I like it... the one word "or" should be "and/or" for being able to use positional arguments and/or keyword arguments.  We can use one, or the other, or both.

Reply:
Added and/or. I'm going to stick with defined as we use `def` as the keywork and https://docs.ruby-lang.org/en/2.3.0/syntax/methods_rdoc.html exclusively uses the language of defining.

Thanks for all that @kotp.